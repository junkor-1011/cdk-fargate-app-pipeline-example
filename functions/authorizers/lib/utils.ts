import type { APIGatewayAuthorizerResult } from 'aws-lambda';

import { decode, JsonWebTokenError, TokenExpiredError, verify, JwtPayload } from 'jsonwebtoken';
import jwks from 'jwks-rsa';
// import { promisify } from 'util';

import { getParameter } from '../lib/aws-utils/ssm';

export const generatePolicy = (
  principal: string,
  effect: 'Allow' | 'Deny',
  resource: string,
): APIGatewayAuthorizerResult => {
  return {
    principalId: principal,
    policyDocument: {
      Version: '2012-10-17',
      Statement: [
        {
          Action: 'execute-api:Invoke',
          Effect: effect,
          Resource: resource,
        },
      ],
    },
  };
};

type VerifiedTokenInfo = JwtPayload & {
  'cognito:username': string;
  'cognito:groups': string[];
  customKey1: string;
  customKey2: string;
  customKey3: string;
};

export const verifyToken = async (token: string): Promise<VerifiedTokenInfo> => {
  // TODO: using environment variable from ssm parameters
  const { stageName } = process.env;
  if (!stageName) {
    throw new Error('failed to get env: stagename');
  }
  const issuer = await getParameter(`/TESTAPP/${stageName}/ISSUER`, true);
  const audience = await getParameter(`/TESTAPP/${stageName}/AUDIENCE`, true);
  const jwksuri = await getParameter(`/TESTAPP/${stageName}/JWKS_URI`, true);

  const decoded = decode(token, { complete: true });
  if (!decoded || !decoded['header'] || !decoded['header'].kid) {
    throw new JsonWebTokenError('invalid token');
  }

  const client = jwks({ jwksUri: jwksuri });
  const key = await client.getSigningKey(decoded['header'].kid);
  const signingKey = key.getPublicKey();

  let res: VerifiedTokenInfo;
  try {
    // TODO: explain why use as(or add user-define type-guard)
    const tokenInfo = verify(token, signingKey, {
      audience,
      issuer,
    }) as VerifiedTokenInfo;
    console.log('tokenInfo: ', tokenInfo);
    res = tokenInfo;
  } catch (err) {
    if (err instanceof TokenExpiredError) {
      throw new Error('token expired');
    }
    if (err instanceof JsonWebTokenError) {
      throw new Error('token is invalid');
    }
    throw err;
  }

  return res;
};
