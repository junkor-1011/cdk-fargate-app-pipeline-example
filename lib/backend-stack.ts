import {
  Aws,
  Duration,
  Stack,
  StackProps,
  aws_iam as iam,
  aws_apigateway as apigateway,
  aws_cognito as cognito,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as ssm from 'aws-cdk-lib/aws-ssm';

type StageStackProps = StackProps & {
  stageName: string;
};

export class BackendStack extends Stack {
  constructor(scope: Construct, id: string, props: StageStackProps) {
    super(scope, id, props);

    const readSSMParamLambdaRole = new iam.Role(this, 'ReadSSMParamLambdaRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      description: 'Role for Lambda reading param from ssm',
    });
    readSSMParamLambdaRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
    );
    readSSMParamLambdaRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMReadOnlyAccess'));

    const pool = new cognito.UserPool(this, `${props.stageName}-Pool`);
    const client = pool.addClient(`${props.stageName}-app-client`, {
      oAuth: {
        flows: {
          authorizationCodeGrant: true,
        },
        scopes: [cognito.OAuthScope.OPENID],
      },
      authFlows: { adminUserPassword: true }, // use cognitoIdp:adminInitiateAuth API
      generateSecret: true,
      refreshTokenValidity: Duration.hours(12),
    });

    const preTokenGenerationLambda = new NodejsFunction(this, 'preTokenGenerationLambda', {
      entry: 'functions/UserPoolTriggers/preTokenGeneration.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
      bundling: {
        sourceMap: true,
      },
      environment: {
        stageName: props.stageName,
      },
    });

    pool.addTrigger(cognito.UserPoolOperation.PRE_TOKEN_GENERATION, preTokenGenerationLambda);

    const helloLambda = new NodejsFunction(this, 'Hello', {
      entry: 'functions/api-backends/hello/get.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(30),
      // role: iamRoleForLambda,
      environment: {
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
        stageName: props.stageName,
      },
      memorySize: 128,
    });

    const acceptTestLambda = new NodejsFunction(this, 'AcceptTest', {
      entry: 'functions/api-backends/lambda-authorizer-test/hello-accept/get.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(30),
      environment: {
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
        stageName: props.stageName,
      },
      memorySize: 128,
      bundling: {
        sourceMap: true,
      },
    });

    const deniedTestLambda = new NodejsFunction(this, 'DeniedTest', {
      entry: 'functions/api-backends/lambda-authorizer-test/hello-denied/get.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(30),
      environment: {
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
        stageName: props.stageName,
      },
      memorySize: 128,
      bundling: {
        sourceMap: true,
      },
    });

    const tokenAuthorizerColorLambda = new NodejsFunction(this, 'ColorAuthorizerLambda', {
      entry: 'functions/authorizers/color/index.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(20),
      role: readSSMParamLambdaRole,
      bundling: {
        sourceMap: true,
      },
      environment: {
        stageName: props.stageName,
      },
    });

    const tokenAuthorizerFruitsLambda = new NodejsFunction(this, 'FruitsAuthorizerLambda', {
      entry: 'functions/authorizers/fruits/index.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(20),
      role: readSSMParamLambdaRole,
      bundling: {
        sourceMap: true,
      },
      environment: {
        stageName: props.stageName,
      },
    });

    const cognitoAuthorizer = new apigateway.CognitoUserPoolsAuthorizer(this, 'cognitoAuthorizer', {
      authorizerName: `${props.stageName}-CognitoAurhorizer`,
      cognitoUserPools: [pool],
    });

    const colorAuthorizer = new apigateway.TokenAuthorizer(this, 'ColorAuthorizer', {
      authorizerName: `${props.stageName}-ColorAurhorizer`,
      handler: tokenAuthorizerColorLambda,
    });

    const fruitsAuthorizer = new apigateway.TokenAuthorizer(this, 'FruitsAuthorizer', {
      authorizerName: `${props.stageName}-FruitsAurhorizer`,
      handler: tokenAuthorizerFruitsLambda,
    });

    const api = new apigateway.RestApi(this, 'helloApigateway', {
      restApiName: `${props.stageName}-testapp-apigateway`,
    });

    const hello = api.root.addResource('hello');
    hello.addMethod('GET', new apigateway.LambdaIntegration(helloLambda), {
      authorizer: cognitoAuthorizer,
    });
    const lambdaAuthorizerTest = api.root.addResource('lambda-authorizer-test');
    lambdaAuthorizerTest
      .addResource('hello-accept')
      .addMethod('GET', new apigateway.LambdaIntegration(acceptTestLambda), {
        authorizer: colorAuthorizer,
      });
    lambdaAuthorizerTest
      .addResource('hello-denied')
      .addMethod('GET', new apigateway.LambdaIntegration(deniedTestLambda), {
        authorizer: fruitsAuthorizer,
      });

    // SSM Parameteres
    new ssm.StringParameter(this, `poolid-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/COGNITO_POOL_ID`,
      stringValue: pool.userPoolId,
      type: ssm.ParameterType.SECURE_STRING,
    });
    new ssm.StringParameter(this, `poolclientid-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/COGNITO_POOL_CLIENT_ID`,
      stringValue: client.userPoolClientId,
      type: ssm.ParameterType.SECURE_STRING,
    });
    new ssm.StringParameter(this, `poolclient-secret-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/COGNITO_POOL_CLIENT_SECRET`,
      stringValue: client.userPoolClientSecret.unsafeUnwrap(),
      type: ssm.ParameterType.SECURE_STRING,
    });

    new ssm.StringParameter(this, `audience-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/AUDIENCE`,
      stringValue: client.userPoolClientId,
      type: ssm.ParameterType.SECURE_STRING,
    });
    new ssm.StringParameter(this, `issuer-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/ISSUER`,
      stringValue: `https://cognito-idp.${Aws.REGION}.amazonaws.com/${pool.userPoolId}`,
      type: ssm.ParameterType.SECURE_STRING,
    });
    new ssm.StringParameter(this, `jwksuri-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/JWKS_URI`,
      stringValue: `https://cognito-idp.${Aws.REGION}.amazonaws.com/${pool.userPoolId}/.well-known/jwks.json`,
      type: ssm.ParameterType.SECURE_STRING,
    });

    new ssm.StringParameter(this, `restapiId-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/REST_API_ID`,
      stringValue: api.restApiId,
      type: ssm.ParameterType.STRING,
    });
    new ssm.StringParameter(this, `restapiUrl-${props.stageName}`, {
      parameterName: `/TESTAPP/${props.stageName}/REST_API_BASE_URL`,
      stringValue: api.url,
      type: ssm.ParameterType.STRING,
    });
  }
}
