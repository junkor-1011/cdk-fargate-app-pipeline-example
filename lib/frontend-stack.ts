import { Stack, StackProps, aws_ecs as ecs } from 'aws-cdk-lib';
import { DockerImageAsset, NetworkMode } from 'aws-cdk-lib/aws-ecr-assets';
import { Construct } from 'constructs';

type StageStackProps = StackProps & {
  stageName: string;
};

export class FrontendStack extends Stack {
  constructor(scope: Construct, id: string, props: StageStackProps) {
    super(scope, id, props);

    const fargateTaskDefinition = new ecs.FargateTaskDefinition(this, `${props.stageName}-TaskDef`, {
      memoryLimitMiB: 2048,
      cpu: 256,
    });
    const asset = new DockerImageAsset(this, 'NextjsAppImage', {
      directory: 'images/nextjs-app',
      networkMode: NetworkMode.HOST,
    });
    const container = fargateTaskDefinition.addContainer('WebContainer', {
      image: ecs.ContainerImage.fromDockerImageAsset(asset),
      // ... other options here ...
    });
  }
}
