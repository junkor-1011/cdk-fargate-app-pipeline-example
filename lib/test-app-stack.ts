import {
  Duration,
  Stack,
  StackProps,
  // aws_iam as iam,
  // aws_apigateway as apigateway,
  aws_cognito as cognito,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

export class TestAppStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const pool = new cognito.UserPool(this, 'Pool');
    pool.addClient('app-client', {
      oAuth: {
        flows: {
          authorizationCodeGrant: true,
        },
        scopes: [cognito.OAuthScope.OPENID],
        // callbackUrls: ['http://localhost:3000/api/auth/callback/cognito'],
      },
      authFlows: { adminUserPassword: true }, // use cognitoIdp:adminInitiateAuth API
      generateSecret: false,
      refreshTokenValidity: Duration.hours(12),
    });
    /*
    pool.addDomain('CognitoDomain', {
      cognitoDomain: {
        // TODO: SHOULD GET VALUE FROM SSM PARAMETER OR SECRETS MANAGER
        domainPrefix: 'super-ultra-hyper-extreme-temporary-app', // TMP
      },
    });
    */

    const preTokenGenerationLambda = new NodejsFunction(this, 'preTokenGenerationLambda', {
      entry: 'functions/UserPoolTriggers/preTokenGeneration.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
    });

    pool.addTrigger(cognito.UserPoolOperation.PRE_TOKEN_GENERATION, preTokenGenerationLambda);

    /*
    const iamRoleForLambda = new iam.Role(this, 'iamRoleForLambda', {
      roleName: `hello-lambda-role`,
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')],
    });

    const helloLambda = new NodejsFunction(this, 'Hello', {
      entry: 'functions/hello/get.ts',
      handler: 'lambdaHandler',
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(30),
      role: iamRoleForLambda,
      environment: {
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
      },
      memorySize: 128,
    });

    const helloApi = new apigateway.RestApi(this, 'helloApigateway', {
      restApiName: `testapp-apigateway`,
    });

    const sample = helloApi.root.addResource('hello');
    const courseSearchIntegration = new apigateway.LambdaIntegration(helloLambda);
    sample.addMethod('GET', courseSearchIntegration);
    */
  }
}
