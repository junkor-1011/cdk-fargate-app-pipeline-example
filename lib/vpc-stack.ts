import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as ssm from 'aws-cdk-lib/aws-ssm';

const stageNames = {
  dev: 'Dev',
  prod: 'Prod',
};

export class VpcStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    const vpcDev = new ec2.Vpc(this, 'TestAppVpc-Dev', {
      cidr: '10.2.0.0/16',
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public-dev',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private-dev',
          subnetType: ec2.SubnetType.PRIVATE_WITH_NAT,
        },
        {
          cidrMask: 24,
          name: 'isolate-dev',
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
    });
    const vpcProd = new ec2.Vpc(this, 'TestAppVpc-Prod', {
      cidr: '10.3.0.0/16',
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public-prod',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private-prod',
          subnetType: ec2.SubnetType.PRIVATE_WITH_NAT,
        },
        {
          cidrMask: 24,
          name: 'isolate-prod',
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
    });

    // SSM ParameterStore
    new ssm.StringParameter(this, 'vpcid-dev', {
      parameterName: `/TESTAPP/${stageNames.dev}/VPC_ID`,
      stringValue: vpcDev.vpcId,
      type: ssm.ParameterType.STRING,
    });
    new ssm.StringParameter(this, 'vpcid-prod', {
      parameterName: `/TESTAPP/${stageNames.prod}/VPC_ID`,
      stringValue: vpcProd.vpcId,
      type: ssm.ParameterType.STRING,
    });
  }
}
