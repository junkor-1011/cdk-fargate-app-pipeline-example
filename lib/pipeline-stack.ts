import { Stack, StackProps, Stage } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as codecommit from 'aws-cdk-lib/aws-codecommit';
import * as pipelines from 'aws-cdk-lib/pipelines';

// import { TestAppStack } from '../lib/test-app-stack';
import { BackendStack } from '../lib/backend-stack';
import { FrontendStack } from '../lib/frontend-stack';

export class PipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // The code that defines your stack goes here

    const repo = new codecommit.Repository(this, 'Repository', {
      repositoryName: 'test-webapp-pipeline-repo',
      description: 'test for cdk pipeline',
    });

    const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
      synth: new pipelines.ShellStep('Synth', {
        input: pipelines.CodePipelineSource.codeCommit(repo, 'main'),
        commands: [
          'yarn install --ignore-scripts',
          'yarn build',
          'yarn cdk synth PipelineStack BackendStack FrontendStack',
        ],
      }),
      codeBuildDefaults: {
        buildEnvironment: {
          privileged: true,
        },
      },
    });
    const devStage = new ApplicationStage(this, 'Dev', {
      stageName: 'dev',
    });
    const prodStage = new ApplicationStage(this, 'Prod', {
      stageName: 'prod',
    });
    pipeline.addStage(devStage);
    pipeline.addStage(prodStage, {
      pre: [new pipelines.ManualApprovalStep('PromoteToProd')],
    });
  }
}

// TODO: Stage VPC
type StageStackProps = StackProps & {
  stageName: string;
};

class ApplicationStage extends Stage {
  constructor(scope: Construct, id: string, props: StageStackProps) {
    super(scope, id, props);

    new BackendStack(this, 'BackendStack', {
      stageName: props.stageName,
    });
    new FrontendStack(this, 'FrontendStack', {
      stageName: props.stageName,
    });
  }
}
