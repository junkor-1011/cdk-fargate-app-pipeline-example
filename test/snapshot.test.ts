import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { BackendStack } from '../lib/backend-stack';
import { FrontendStack } from '../lib/frontend-stack';
import { PipelineStack } from '../lib/pipeline-stack';

describe('snapshot test', () => {
  it("PipelineStack's Snapshot test", () => {
    const app = new cdk.App();
    const stack = new PipelineStack(app, 'TestAppStack');
    const template = Template.fromStack(stack).toJSON();

    expect(template).toMatchSnapshot();
  });
  it("BackendStack's Snapshot test", () => {
    const app = new cdk.App();
    const stack = new BackendStack(app, 'BackendStack', {
      stageName: 'test',
    });
    const template = Template.fromStack(stack).toJSON();

    expect(template).toMatchSnapshot();
  });
  it("FrontendStack's Snapshot test", () => {
    const app = new cdk.App();
    const stack = new FrontendStack(app, 'FrontendStack', {
      stageName: 'test',
    });
    const template = Template.fromStack(stack).toJSON();

    expect(template).toMatchSnapshot();
  });
});
