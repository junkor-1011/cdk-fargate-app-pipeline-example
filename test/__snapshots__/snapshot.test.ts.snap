// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test TestApp Stack's Snapshot test 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "PoolD3F588B8": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccountRecoverySetting": Object {
          "RecoveryMechanisms": Array [
            Object {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            Object {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": Object {
          "AllowAdminCreateUserOnly": true,
        },
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "LambdaConfig": Object {
          "PreTokenGeneration": Object {
            "Fn::GetAtt": Array [
              "preTokenGenerationLambda1F130400",
              "Arn",
            ],
          },
        },
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "VerificationMessageTemplate": Object {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Retain",
    },
    "Poolappclient2D3B8A6A": Object {
      "Properties": Object {
        "AllowedOAuthFlows": Array [
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": Array [
          "openid",
        ],
        "CallbackURLs": Array [
          "https://example.com",
        ],
        "ExplicitAuthFlows": Array [
          "ALLOW_ADMIN_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "GenerateSecret": false,
        "RefreshTokenValidity": 720,
        "SupportedIdentityProviders": Array [
          "COGNITO",
        ],
        "TokenValidityUnits": Object {
          "RefreshToken": "minutes",
        },
        "UserPoolId": Object {
          "Ref": "PoolD3F588B8",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "preTokenGenerationLambda1F130400": Object {
      "DependsOn": Array [
        "preTokenGenerationLambdaServiceRole0C3B4FA0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a7be855f0aa7f8abe4a378fb16c3a14b0a1800ea6800434bdf9afab7fe6838cc.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.lambdaHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "preTokenGenerationLambdaServiceRole0C3B4FA0",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "preTokenGenerationLambdaPreTokenGenerationCognitoB0148D6A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "preTokenGenerationLambda1F130400",
            "Arn",
          ],
        },
        "Principal": "cognito-idp.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "PoolD3F588B8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "preTokenGenerationLambdaServiceRole0C3B4FA0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
